edb/common/checked.py

```diff
@@ -85,14 +82,29 @@ def __class_getitem__(
                 raise TypeError(f"{cls!r} expects two type parameters")
             type_dict["keytype"] = params[0]
             type_dict["valuetype"] = params[1]
+        if not all(isinstance(param, type) for param in params):
+            if not all(type(param) == TypeVar for param in params):
+                raise TypeError(f"{cls!r} expects types as type parameters")
+            # All parameters are type variables: return the regular generic
+            # alias to allow proper subclassing.
+            generic = super(ParametricType, cls)
+            return generic.__class_getitem__(params)  # type: ignore
+
         return type(name, bases, type_dict)
 
     def __reduce__(self) -> Tuple[Any, ...]:
         assert self.types is not None
-        base: Type[ParametricType] = self.__class__.__bases__[0]
+        cls: Type[ParametricType] = self.__class__
         container = getattr(self, "_container", ())
+        if cls.__name__.endswith("]"):
```

Comment:
We're using `typing_inspect` module by Ivan (https://github.com/ilevkivskyi/typing_inspect/blob/master/typing_inspect.py) -- can we perhaps write a cleaner check here using his APIs or typing APIs?

Reply:
`typing_inspect` will not work: the runtime type checking facility required for us to use a custom `ParametricType` base class which provides `__class_getitem__` that does exactly what we want: constructs a real subclass on the fly with the parametrized types applied.

Since it's a real subclass (unlike what `Generic[T]` does which is to give you an *instance* of `GenericAlias` which itself is not a real type), it has to have a unique name. So we give it a name that includes the generics.

Checking if we're dealing with such dynamic subclass by looking at the last character of the name is maybe not the most elegant thing in the entire module but it's the easiest solution for what we need.

For some context, the reason for the check under which you commented is to identify if we're pickling an instance of this dynamically generated type. In such case, we need to use `__restore__` for unpickling to recreate this dynamic type just-in-time.

If we're dealing with a type that's not the dynamically generated one, that means we are dealing with a subclass of it. If that subclass is seen from global state, we don't have to do anything special, the dynamic base type is already present and properly parametrized by the time our subclass is created.

Reply:
Please copy/paste this comment into code :)